File: Java/Concurrency/CustomThreadPoolDemo.java
package Concurrency;

import java.util.LinkedList;
import java.util.Queue;

public class CustomThreadPoolDemo {

    static class CustomThreadPool {
        private final Queue<Runnable> taskQueue = new LinkedList<>();
        private final Worker[] workers;
        private final Object lock = new Object();
        private volatile boolean isShutdown = false;

        CustomThreadPool(int numThreads) {
            workers = new Worker[numThreads];
            for (int i = 0; i < numThreads; i++) {
                workers[i] = new Worker("Worker-" + i);
                workers[i].start();
            }
        }

        public void submit(Runnable task) {
            synchronized (lock) {
                if (isShutdown) {
                    throw new IllegalStateException("ThreadPool is shut down, cannot accept new tasks");
                }
                taskQueue.add(task);
                lock.notify();
            }
        }

        public void shutdown() {
            synchronized (lock) {
                isShutdown = true;
                lock.notifyAll();
            }
        }

        public void awaitTermination() throws InterruptedException {
            for (Worker worker : workers) {
                worker.join();
            }
        }

        private class Worker extends Thread {
            Worker(String name) {
                super(name);
            }

            @Override
            public void run() {
                while (true) {
                    Runnable task;
                    synchronized (lock) {
                        while (taskQueue.isEmpty() && !isShutdown) {
                            try {
                                lock.wait(); //worker need to wait
                            } catch (InterruptedException e) {
                                Thread.currentThread().interrupt();
                                return;
                            }
                        }
                        if (taskQueue.isEmpty() && isShutdown) {
                            return; // no more work, pool is shutting down
                        }

                        // worker will pull the task and process the task
                        task = taskQueue.poll();
                    }
                    try {
                        task.run();
                    } catch (RuntimeException e) {
                        System.out.println(getName() + " task threw exception: " + e.getMessage());
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        CustomThreadPool pool = new CustomThreadPool(3);

        for (int i = 0; i < 10; i++) {
            final int taskId = i;
            pool.submit(() -> {
                System.out.println(Thread.currentThread().getName() + " executing task " + taskId);
                try {
                    Thread.sleep(200);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
        }

        pool.shutdown();
        pool.awaitTermination();
        System.out.println("All tasks completed, pool terminated.");
    }
}

File: Python/Concurrency/CustomThreadPoolDemo.py
import threading
import time
from collections import deque
from typing import Callable, Deque


class CustomThreadPool:
    def __init__(self, num_threads: int) -> None:
        self._task_queue: Deque[Callable[[], None]] = deque()
        self._lock = threading.Lock()
        self._not_empty = threading.Condition(self._lock)
        self._is_shutdown = False
        self._workers = [
            threading.Thread(target=self._work, name=f"Worker-{i}")
            for i in range(num_threads)
        ]
        for worker in self._workers:
            worker.start()

    def submit(self, task: Callable[[], None]) -> None:
        with self._not_empty:
            if self._is_shutdown:
                raise RuntimeError("ThreadPool is shut down, cannot accept new tasks")
            self._task_queue.append(task)
            self._not_empty.notify()

    def shutdown(self) -> None:
        with self._not_empty:
            self._is_shutdown = True
            self._not_empty.notify_all()

    def await_termination(self) -> None:
        for worker in self._workers:
            worker.join()

    def _work(self) -> None:
        while True:
            with self._not_empty:
                while not self._task_queue and not self._is_shutdown:
                    self._not_empty.wait()
                if not self._task_queue and self._is_shutdown:
                    return  # no more work, pool is shutting down
                task = self._task_queue.popleft()

            try:
                task()
            except Exception as e:  # noqa: BLE001
                print(f"{threading.current_thread().name} task threw exception: {e}")


def main() -> None:
    pool = CustomThreadPool(3)

    def make_task(task_id: int) -> Callable[[], None]:
        def task() -> None:
            print(f"{threading.current_thread().name} executing task {task_id}")
            time.sleep(0.2)
        return task

    for i in range(10):
        pool.submit(make_task(i))

    pool.shutdown()
    pool.await_termination()
    print("All tasks completed, pool terminated.")


if __name__ == "__main__":
    main()
